### DISCUSSION TITLE
0.96" TFT 160x80 marked ST7735 (but actually GC9106) cropped display - SOLVED

### DISCUSSION THREAD

**Discussion by Boyeen** [Bodmer/TFT_eSPI#1310] [Compatible displays and setup files]
I see a lot of Chinese vendors now selling these displays for cheap which appear to be defective.... they have a blue label and under TFT_eSPI, behave like 120x80 displays.... the screen is black at the end furthest from the ribbon. I have been looking into this and have a partial solution but am a babe-in-the-woods on LCD tech. Hopefully someone can help.

First, full disclosure, I got GREAT help from Dave Prentice on the Arduino forum, could not have gotten this far alone.
I obtained an 8051 driver from the Chinese vendor and Dave converted it and made a simple test sketch which works, it can
access the entire display (80x160) in simple fillRect() drawRect() style. IMHO the problem with these displays seems to be large screen offsets but ALSO that the controller is a clone of ST7735S, the init() code uses several undeclared controller registers/addresses.

My application is a jpg image load of the display in portrait mode. My code works with my "GOOD" displays (ST7735REDTAB). Of course I thought a bunch of $2 displays was too good an opportunity to pass up. More fool me?

I took the init() routine of Daves code and hacked it into ST7735_init(define)(rotate) driver files by adding a temporary INITR_BLUETAB #define. It's working in that I can access the whole display height (160) but my images are offset in the width (80) dimension. As I said, the 8051 code has correct geometry, so perhaps I am abusing Bodmers masterpiece? :-)

![IMG_1291](https://user-images.githubusercontent.com/6819619/131193598-75331dbf-18b9-43cc-9425-c696e2514ba0.jpg)

Sorry if I am not using correct LCD terminology... I know these are not simple bitmapped displays, but this is all new to me.

Q1. Where do I start trying to correct the geometry?
Q2. Is there some documentation on how to add/extend a driver?

I am happy to share my code, especially when I get it to work... but this msg has been long enough :-)

Here is the init() code I am using in st7735_init.h:

` if (tabcolor == INITB)
{
commandList(Bcmd);
}else if(tabcolor == INITR_BLUETAB) {

   Serial.println("Initializing a blue display");
writecommand(0xfe);
writecommand(0xfe);
writecommand(0xfe);
writecommand(0xef);

writecommand(0xb3);
writedata(0x03);

writecommand(0x36);
writedata(0xd8);

writecommand(0x3a);
writedata(0x05);

writecommand(0xb6);
writedata(0x11);
writecommand(0xac);
writedata(0x0b);

writecommand(0xb4);
writedata(0x21);

writecommand(0xb1);
writedata(0xc0);

writecommand(0xe6);
writedata(0x50);
writedata(0x43);
writecommand(0xe7);
writedata(0x56);
writedata(0x43);

writecommand(0xF0);
writedata(0x1f);
writedata(0x41);
writedata(0x1B);
writedata(0x55);
writedata(0x36);
writedata(0x3d);
writedata(0x3e);
writedata(0x0);
writedata(0x16);
writedata(0x08);
writedata(0x09);
writedata(0x15);
writedata(0x14);
writedata(0xf);

writecommand(0xF1);

writedata(0x1f);
writedata(0x41);
writedata(0x1B);
writedata(0x55);
writedata(0x36);
writedata(0x3d);
writedata(0x3e);
writedata(0x0);
writedata(0x16);
writedata(0x08);
writedata(0x09);
writedata(0x15);
writedata(0x14);
writedata(0xf);

writecommand(0xfe);
writecommand(0xff);

writecommand(0x35);
writedata(0x00);
writecommand(0x44);
writedata(0x00);
writecommand(0x11);
delay (120);
writecommand(0x29);
//writecommand(0x2c);

writecommand(0x2A); //Set Column Address
writedata(0x00);
writedata(0x18);
writedata(0x00);
writedata(0x00);//(0x67);
writecommand(0x2B); //Set Page Address
writedata(0x00);
writedata(0x00);
writedata(0x00);
writedata(0x9f);
writecommand(0x2c);
// writecommand(TFT_INVON);
colstart = 1;
rowstart = 26;
}`

**Comment by Boyeen** [comment]
If anyone has one of these displays that you thought was broken, here is a working sketch based on an 8051 driver example I received from the Chinese vendor. Perhaps it will help get someone started...  I took the function **initial()** from this sketch and added it to Bodmers ST7735_init.h for my "BLUETAB" version which is the subject of this topic.

`

#include <SPI.h>

#define CS  10
#define DC  9
#define RST 8
#define BL  7

#define uchar unsigned char
#define uint  unsigned int
#define Xoffset 24
#define Yoffset 0

#define RED         0xf800
#define GREEN       0x07e0
#define BLUE        0x001f
#define WHITE       0xffff
#define BLACK       0x0000
#define YELLOW  0xFFE0
#define GRAY0   0xEF7D
#define GRAY1   0x8410
#define GRAY2   0x4208


void  Contrast_Adjust();

//code uint typFNT_GBK16 hz16[];

void Output_Pixel(uint x, uint y);

void delay_ms(uint time)
{
    delay(time);
}

void  SPI_WriteData(uchar Data)
{
    SPI.transfer(Data);
}

void  Lcd_WriteIndex(uchar Data)
{
    digitalWrite(CS, LOW);
    digitalWrite(DC, LOW);
    SPI_WriteData(Data);
    digitalWrite(CS, HIGH);
}
//ÏòÒº¾§ÆÁÐ´Ò»¸ö8Î»Êý¾Ý
void  Lcd_WriteData(uchar Data)
{
    digitalWrite(CS, LOW);
    digitalWrite(DC, HIGH);
    SPI_WriteData(Data);
    digitalWrite(CS, HIGH);
}
//ÏòÒº¾§ÆÁÐ´Ò»¸ö16Î»Êý¾Ý
void  LCD_WriteData_16Bit(unsigned int Data)
{
    digitalWrite(CS, LOW);
    digitalWrite(DC, HIGH);
    SPI_WriteData(Data >> 8);   //Ð´Èë¸ß8Î»Êý¾Ý
    SPI_WriteData(Data);            //Ð´ÈëµÍ8Î»Êý¾Ý
    digitalWrite(CS, HIGH);
}

void Reset()
{
    digitalWrite(RST, LOW);
    delay_ms(100);
    digitalWrite(RST, HIGH);
    delay_ms(100);
}

//////////////////////////////////////////////////////////////////////////////////////////////
//Òº¾§ÆÁ³õÊ¼»¯
void initial(void)
{

    Lcd_WriteIndex(0xfe);
    Lcd_WriteIndex(0xfe);
    Lcd_WriteIndex(0xfe);
    Lcd_WriteIndex(0xef);

    Lcd_WriteIndex(0xb3);
    Lcd_WriteData(0x03);

    Lcd_WriteIndex(0x36);
    Lcd_WriteData(0xd8);

    Lcd_WriteIndex(0x3a);
    Lcd_WriteData(0x05);

    Lcd_WriteIndex(0xb6);
    Lcd_WriteData(0x11);
    Lcd_WriteIndex(0xac);
    Lcd_WriteData(0x0b);

    Lcd_WriteIndex(0xb4);
    Lcd_WriteData(0x21);

    Lcd_WriteIndex(0xb1);
    Lcd_WriteData(0xc0);

    Lcd_WriteIndex(0xe6);
    Lcd_WriteData(0x50);
    Lcd_WriteData(0x43);
    Lcd_WriteIndex(0xe7);
    Lcd_WriteData(0x56);
    Lcd_WriteData(0x43);

    Lcd_WriteIndex(0xF0);
    Lcd_WriteData(0x1f);
    Lcd_WriteData(0x41);
    Lcd_WriteData(0x1B);
    Lcd_WriteData(0x55);
    Lcd_WriteData(0x36);
    Lcd_WriteData(0x3d);
    Lcd_WriteData(0x3e);
    Lcd_WriteData(0x0);
    Lcd_WriteData(0x16);
    Lcd_WriteData(0x08);
    Lcd_WriteData(0x09);
    Lcd_WriteData(0x15);
    Lcd_WriteData(0x14);
    Lcd_WriteData(0xf);

    Lcd_WriteIndex(0xF1);

    Lcd_WriteData(0x1f);
    Lcd_WriteData(0x41);
    Lcd_WriteData(0x1B);
    Lcd_WriteData(0x55);
    Lcd_WriteData(0x36);
    Lcd_WriteData(0x3d);
    Lcd_WriteData(0x3e);
    Lcd_WriteData(0x0);
    Lcd_WriteData(0x16);
    Lcd_WriteData(0x08);
    Lcd_WriteData(0x09);
    Lcd_WriteData(0x15);
    Lcd_WriteData(0x14);
    Lcd_WriteData(0xf);

    Lcd_WriteIndex(0xfe);
    Lcd_WriteIndex(0xff);

    Lcd_WriteIndex(0x35);
    Lcd_WriteData(0x00);
    Lcd_WriteIndex(0x44);
    Lcd_WriteData(0x00);
    Lcd_WriteIndex(0x11);
    delay_ms (120);
    Lcd_WriteIndex(0x29);
    //Lcd_WriteIndex(0x2c);

    Lcd_WriteIndex(0x2A); //Set Column Address
    Lcd_WriteData(0x00);
    Lcd_WriteData(0x18);
    Lcd_WriteData(0x00);
    Lcd_WriteData(0x67);
    Lcd_WriteIndex(0x2B); //Set Page Address
    Lcd_WriteData(0x00);
    Lcd_WriteData(0x00);
    Lcd_WriteData(0x00);
    Lcd_WriteData(0x9f);
    Lcd_WriteIndex(0x2c);
}

void Lcd_SetRegion(unsigned int x_start, unsigned int y_start, unsigned int x_end, unsigned int y_end)
{

    Lcd_WriteIndex(0x2a);
    Lcd_WriteData(0x00);
    Lcd_WriteData(x_start + Xoffset);
    Lcd_WriteData(0x00);
    Lcd_WriteData(x_end + Xoffset);

    Lcd_WriteIndex(0x2b);
    Lcd_WriteData(0x00);
    Lcd_WriteData(y_start + Yoffset);
    Lcd_WriteData(0x00);
    Lcd_WriteData(y_end + Yoffset);
    Lcd_WriteIndex(0x2c);
}

void PutPixel(uint x_start, uint y_start, uint color)
{
    Lcd_SetRegion(x_start, y_start, x_start + 1, y_start + 1);
    LCD_WriteData_16Bit(color);
}

void fillRect(uint x, uint y, uint w, uint h, uint color)
{
    Lcd_SetRegion(x, y, x + w - 1, y + h - 1);
    for (int row = 0; row < h; row++) {
        for (int col = 0; col < w; col++) {
            LCD_WriteData_16Bit(color);
        }
    }
}

void drawRect(uint x, uint y, uint w, uint h, uint color)
{
    fillRect(x, y,         w, 1, color);
    fillRect(x, y + h - 1, w, 1, color);
    fillRect(x,         y, 1, h, color);
    fillRect(x + w - 1, y, 1, h, color);
}

void dsp_single_colour(int color)
{
    fillRect(0, 0, 80, 160, color);
    drawRect(0, 0, 80, 160, WHITE); //outer rectangle
    drawRect(8, 8, 64, 144, BLACK); //inner rectangle
    delay(1000);
}

void setup()
{
    SPI.begin();  // ADD Your MOSI and SCK pin arguments here if necessary
    pinMode(CS, OUTPUT);
    pinMode(DC, OUTPUT);
    pinMode(RST, OUTPUT);
    pinMode(BL, OUTPUT);
    digitalWrite(BL, HIGH);
    Reset();
    initial();
}

void loop(void)
{
    //Font_Test();
    dsp_single_colour(0xf800);//RED
    dsp_single_colour(0x07e0);//GREEN
    dsp_single_colour(0x001f);//BLUE
    delay_ms(5000);
}

#if 0
{
    // my style of init table: (cmd), N, data... 
    (0xfe), 0,
    (0xfe), 0,
    (0xfe), 0,
    (0xef), 0,
    (0xb3), 1, 0x03,
    (0x36), 1, 0xd8,
    (0x3a), 1, 0x05,
    (0xb6), 1, 0x11,
    (0xac), 1, 0x0b,
    (0xb4), 1, 0x21,
    (0xb1), 1, 0xc0,
    (0xe6), 2, 0x50, 0x43,
    (0xe7), 2, 0x56, 0x43,
    (0xF0), 12, 0x1f, 0x41, 0x1B, 0x55, 0x36, 0x3d, 0x3e, 0x0, 0x16, 0x08, 0x09, 0x15, 0x14, 0xf,
    (0xF1), 12, 0x1f, 0x41, 0x1B, 0x55, 0x36, 0x3d, 0x3e, 0x0, 0x16, 0x08, 0x09, 0x15, 0x14, 0xf,
    (0xfe), 0,
    (0xff), 0,
    (0x35), 1, 0x00,
    (0x44), 1, 0x00,
    (0x11), 0,
    delay_ms (120);
    (0x29), 0,
    /*** (0x2c), 0,***/
    (0x2A), 4, /***Set Column Address***/ 0x00, 0x18, 0x00, 0x67,
    (0x2B), 4, /***Set Page Address***/ 0x00, 0x00, 0x00, 0x9f,
    (0x2c), 0,
};
#endif
`

And here is the output:

![IMG_1288](https://user-images.githubusercontent.com/6819619/131201379-3f4fd2b8-ace9-4452-bd54-9d6199a78ba1.jpg)

**Comment by Boyeen** [comment]
I may be talking to myself here but in case it is of use to someone in the future, I found an error I made in hacking Bodmers code:

In ST7735_Defines.h ...    For my "new" BLUE TAB check I forgot to define CGRAM OFFSET.

So now I am closer in all orientations but the remaining offsets, though small are not acceptable...  suggestions welcome :-)

![IMG_1296](https://user-images.githubusercontent.com/6819619/131232366-c5d06048-2c79-43ba-8583-87d1fb716799.jpg)
![IMG_1297](https://user-images.githubusercontent.com/6819619/131232367-fcbe8592-dc0f-482d-80f8-34ce3df726b6.jpg)
![IMG_1298](https://user-images.githubusercontent.com/6819619/131232374-7669a70b-0d31-400a-892c-929b689668e9.jpg)
![IMG_1299](https://user-images.githubusercontent.com/6819619/131232376-69e688a4-0dc5-4736-842d-fdb6e7124257.jpg)

**Reply by Sicco55** [reply]
Hi, thanks for your good work. I have also some of these displays. My question: how do you get the text on the display?

**Reply by Boyeen** [reply]
It;s just the same as any display. Once you have a driver specified, any demo code should work.

**Reply by Sicco55** [reply]
Thanks for your answer but for me it is not clear how to do this. As in any display I use e.g:  ST7789 display, Arduino_ST7789 tft = Arduino_ST7789(etc....)  After that I use: tft.println("Hello World!"); But in your code this command is not usable. The //Font_Test(); declaration is commented out and the function is not there. Can you give a snippet of the code how you have included this? That would be of great help. At least, printing characters is a first thing to make this display usable. Thanks for your work!

**Reply by Boyeen** [reply]
I have a test fork of the library with MY custom support for these display with GC9106 driver chip. This is not official, unsupported, use at your own risk etc but it is working for me so far in several projects.

 https://github.com/Boyeen/TFT_eSPI

You will need to edit  "User_Setup_Select.h" to enable just my custom setup file:    #include <User_Setups/Setup300_GC9106.h> 

If you look in the examples folder:    examples/160x80/gc9106_demo.ino

This is the sample Arduino sketch I was using that tested the 4 orientations of the display. I didn't write it, I got it from David Prentice over on the Arduino board. It should show you how to add text in the basic tft_eSPI library manner.
I'm not an expert but will help if I can, just ask.

**Reply by Sicco55** [reply]
Hi Boyeen. Thanks to my seller the problem is solved. Change in Adafruit-ST7735-Library : Adafruit_ST7xx.h -> #define ST77XX_MADCTL_MX 0x36 // was 0x40. This solved the problem! I use: tft.initR(INITR_MINI160x80);. Now I can use the full graphicstest.ino for this blue tab board.  Look in Sitronics ST7735S V1.3 pg 142.

**Reply by Sicco55** [reply]
Hi Boyeen, thanks for updating TFT_eSPI! The sketch works. Again, thanks for the good work. I am impressed.

**Comment by Boyeen** [comment]
OK finally done. Again, I know I'm talking to myself entirely on this one but someone else may benefit in the future.
I plan more verification and testing but I will make the changes available.  The 8051 init() above works within the
framework of the ST7735 driver.

**Comment by Boyeen** [comment]
Turns out the controller on the display is NOT an ST7735S as is printed on the PCB.... it's a GalaxyCore GC9106.
I guess I will have the first driver for that chip/display combo... :-)

**Reply by gkr321** [reply]
How did you determine that? I don't see anything that says that on mine.

**Comment by Boyeen** [comment]
I’ll be happy to help but will need a little time, I’ll try to work something up later today.

Tim

**Comment by MacSass** [comment]
Hello @Boyeen,
I read the thread here and your discussion here: https://forum.arduino.cc/t/cropped-80x160-display-st7735s-tft-espi/899248/26 
I do have the same display obviously (2 pieces - blue tab) and can´t get them to work with TFT_eSPI.

I must say I have way less knowledge and experience with TFT_eSPI so I´m not sure what I need to change where to get this display working. Currently - whatever I try - I do get colors wrong (Red shows as yellow, green as magenta and blue as cyan), white and black are inverted and I do have an image offset as well ...

Would you be able to provide some information / files how you got your displays to work?
Did you create a "bluetab" setup-file that I might be able to use?

thanks - help would really be appreciated, as those are the smalles LCD I could find so far (and I need small ones) - MacSass

**Reply by MacSass** [reply]
OK, I worked my way through that stuff and have been able to run the "chinese ino" successfully - prooving that my display as such is OK.

I tried modifying ST7735_Defines.h and ST7735_Init.h with the respective sections for a "bluetab" - but I have no clue how to fill the commands with the information from the chinese .ino - can´t figure out how they relate to the commands in the TFT-eSPI driver files.

Would you mind sharing your resulting files?

**Reply by Boyeen** [reply]
No problem. I forked the library here: https://github.com/Boyeen/TFT_eSPI
You can grab my files from there: I think these are what you need… 
1. User_Setup300GC9106.h in User_Setups
2. User_Setup.h and User_Setup_Select.h
2. 3 files in TFT_Drivers: GC9106_XXX.h
3. Optionally a demo in examples:  gc9106_demo.ino

**Reply by MacSass** [reply]
Great - thanks a lot.
Actually I got it to work, but I will check out your files to see if they are less "hacked" than mine :-)
Eventually it might make sense to get that into the official version, or even as a sep. driver? Others might run into the same issue as those displays are cheap and highly available.
Thanks again for sharing your results!

**Reply by Boyeen** [reply]
I agree about offering it to Bodmer to merge… but I am the only person who has tried them so I didn’t feel comfortable. Plus, he would need a display to verify them…. Anyway, glad you got it to work.

**Reply by pruttelherrie** [reply]
> No problem. I forked the library here: https://github.com/Boyeen/TFT_eSPI You can grab my files from there: I think these are what you need…
> 
>     1. User_Setup300GC9106.h in User_Setups
>     2. User_Setup.h and User_Setup_Select.h
>     3. 3 files in TFT_Drivers: GC9106_XXX.h
>     4. Optionally a demo in examples:  gc9106_demo.ino

Anyone following this, please note that you'll have to update TFT_eSPI.cpp as well, and add the defines in the lists:
#elif defined (GC9106_DRIVER)
     #include "TFT_Drivers/GC9106_Init.h"
#elif defined (GC9106_DRIVER)
     #include "TFT_Drivers/GC9106_Rotation.h"

 I got it to work nicely that way.

**Reply by nachus001** [reply]
Hello Boyeen
I'm using your library, great rpoduct! I came through one of those 0.96"tft's labelled ST7735S. I could implement the mod for the GC9106 controller (patched the 2.4 library with them) and actually make the thing run.
But... The display is mirrored and I cannot find a display command that can invert the video accessing in order to un-mirror what is seen in the screen.

I think this is due to how this particular manufacturer wired the panel to the chip.

Other issues, (when running GC9106_demo.ino) I don't know if they're related, but that background colors are not what the text says. "RED" is blue, and "BLUE" is red. However Green and gray are okey.

Would you tell me where in the library to hack in order to  un-mirror the TFT output?

On the other side, i've checked the GC9106_init.h file and found that there are many commands that are not part of the GC9106 command set (nor the ST7735S controller either), and are apparently ignored (I commented them out with no adverse effect)
There are also incomplete commands, two 0xf0 and 0xf1 that require 15 parameters (by datasheet) and are given 14, and one final 0x2C that has a variable amount of parameters (per datasheet) and is issued with no parameters at all

Would you tell me where in the library to hack in order to  un-mirror the TFT output?

thanks in advance
Nachus

**Comment by Bodmer** [comment]
Hi, Missed this discussion, looks like you have it cracked though.

If this display is still available I will accept a pull request. Changes must be the minimum required to support the display with no other edits.

**Reply by Boyeen** [reply]
I would be honored to help the library any way I can Bodmer but I don't think my driver is worthy. I based it on an 8051 driver I got from the Chinese vendor and it is not even "true" to the driver chip (I THINK they use GC9106). It works but...

There are lots of these "lying" boards being delivered from China. They say ST7735 on the silk-screen but in fact use the Chinese part

However, if you think it is worth a try, I will be delighted to make the effort to clean up my code and submit it.

Finally, let me say thanks for your phenomenal efforts. I use so many of your libraries and your example code. I owe you multiple beers :-)

Tim

**Comment by MacSass** [comment]
Hi,
unfortunately I don´t feel competent enough to create a working pull request - hoping that @Boyeen would be able / find the time to do so.
I would be happy to test any beta version of it with my displays, if that helps.
Those displays are still available for low price at Ali-Express, I´d be happy to ship one around in Europe if needed ...

**Comment by Boyeen** [comment]
Hiya Nachus,

I can’t take much credit for the GC9106 mods. I was (like you) just trying to make use of some displays I got from China. In fact I asked the vendor for help and they sent me some 8051 driver code and I converted it over (with help from another user). I never looked at the instruction set so it doesn’t surprise me that the code is weird.

Anyway, from that intro you can probably tell that I don’t think I can be much help to you here. If the colors are wrong, they are probably inverted and there are some commands to handle that. Look in UserSetup.h, it says uncomment one of these:

// #define TFT_INVERSION_ON

// #define TFT_INVERSION_OFF
Sounds like you ALSO would like to mirror the display and I don’t know how to do that. If you want my help (for what it’s worth) to look further send me more info about what you are seeing. Or send me your code and I’ll see if I can spot anything.

Bodmer asked me to request him to pull my mods but I am not sure I trust them.
The mods worked for my displays/projects but I can’t vouch for them really.

Sorry not to be much help.

Tim

**Reply by nachus001** [reply]
Hello Tim
I mnage to solve it, I had to modify code in GC9106_rotation.h
The problem was that the chip in my display was wired differently to the tft panel than yours.
For further reference, here's the code modified.
`
  rotation = m % 4;

  writecommand(TFT_MADCTL);
  Serial.printf("ROT: %d",rotation);
  switch (rotation) {
    case 0: // Portrait
       
      // writedata(TFT_MAD_BGR); original
       writedata(TFT_MAD_BGR | TFT_MAD_MY);
       _width  = _init_width;
       _height = _init_height;
       colstart = 24;
       rowstart = 0;
      break;
    case 1: // Landscape (Portrait + 90)
      //writedata(TFT_MAD_MX | TFT_MAD_MV | TFT_MAD_BGR);
      writedata( TFT_MAD_MV | TFT_MAD_BGR);
      _width  = _init_height;
      _height = _init_width;
       colstart = 0;
       rowstart = 24;

      break;
    case 2: // Inverter portrait
      //writedata(TFT_MAD_MX | TFT_MAD_MY | TFT_MAD_BGR);
      writedata(TFT_MAD_MX | TFT_MAD_BGR);
      _width  = _init_width;
      _height = _init_height;
       colstart = 24;
       rowstart = 0;
      break;
    case 3: // Inverted landscape
      //writedata(TFT_MAD_MV | TFT_MAD_MY | TFT_MAD_BGR);
      writedata( TFT_MAD_MV |  TFT_MAD_MY| TFT_MAD_MX | TFT_MAD_BGR);
      _width  = _init_height;
      _height = _init_width;
       rowstart = 24;
       colstart = 0;
      break;
  }`

Thank You
nachus

**Reply by nachus001** [reply]
Oh, and, in regards of color, the GC9106_defines.h lacked the color order flag. It was hardcoded, and, it worked for your display and not in mine :<
Now, the configuration from platformio.ini works ok. I uploaded it to a repository for anyone who wants to run this display
https://github.com/nachus001/GC9106_driver

Regards
Nachus

**Comment by Boyeen** [comment]
Sounds great Nachus,

I will take a look at the files and try to update mine. Then, if you are willing, I might ask you to download a fresh copy and check if it still works with your display? That will give us 2 points of reference and perhaps we can send a pull request to Bodmer. I will be sure to add you to the credits :-)

Thanks,
Tim

**Reply by nachus001** [reply]
Of course

**Reply by MacSass** [reply]
I´m happy to try any new version with my display as well - unfortunately I don´t feel competent to create a pull request.

As said - if you create an updated version, I´m happy to to try it and provide feedback.

**Comment by rtopz** [comment]
Thanks very much for this driver and the explanation. It worked for my generic chinese display that was also supposed to be st7735.

**Reply by nachus001** [reply]
Your welcomed!

**Comment by Jorre05** [comment]
Hi,
Thanks for al this info. I have the same fake st7735 display which is a gc9106 in reality.
I took the liberty to transform your code into an Esphome component. see https://github.com/Jorre05/gc9106

**Comment by Boyeen** [comment]
As I stated earlier in this thread, the marking on the display are wrong wrt the controller, not uncommon with cheap Chinese displays. That is why it does not work with the base ST7735  library drivers.

I got help from Dave Prentice on the Arduino forum who knows far more than I ever will. Anyway, he had me run some code on the controller chip in my display and he was able to determine from the returned values that it reports as GC9106.

**Reply by Zemin-Lin** [reply]
Thank you for sharing and solving my recent problems. This screen is really annoying.
